src/libcore/iter/traits/iterator.rs

```diff
@@ -388,6 +392,26 @@ pub trait Iterator {
     /// [`once`] is commonly used to adapt a single value into a chain of
     /// other kinds of iteration.
     ///
+    ///
+    /// # Overflowing behavior for long iterators
+    ///
+    /// This method allows to easily build an iterator that yields more than
+    /// [`usize::MAX`] items. In that case, some methods that return `usize`
+    /// are not guarded against overflow. For example, this includes the
+    /// following methods:
+    ///
+    /// - [`Iterator::count`]
+    /// - [`Iterator::enumerate`]
+    /// - [`Iterator::position`] and [`Iterator::rposition`]
+    /// - [`ExactSizeIterator::len`]
+    ///
+    /// An overflow in those methods leads to a wrong result or a panic. If
+    /// debug assertions are enabled, a panic is guaranteed.
```

Comment:
Is it possible that we could guarantee a panic whether or not debug assertions are enabled? Reserving the right to give a wrong result doesn't seem helpful here.

Reply:
Reserving this right is "helpful" in the sense that it means we don't have any performance overhead from overflow checks.

Reply:
It seems wrong that `len()` could return something smaller than `size_hint().0` on an iterator that implements TrustedLen. What sort of overhead are we talking about in practice?

Reply:
I was referring to all methods listed in that doc comment, not just `len`. `Iterator::count` currently does not check for overflow in release builds because it simply adds 1 repeatedly. In the case of `len`, I agree that it should probably panic rather than return an incorrect value, if we were to land this change.

Reply:
`len` guarantees a panic on overflow. I changed this in response to [this comment](https://github.com/rust-lang/rust/pull/66531#discussion_r348002730). 

This sentence here is a bit vague, yes. It's still technically correct for `len`, but maybe I should add something like "`ExactIterator::len` even guarantees a panic in release mode"? On the other hand, this behavior is already documented on that method (see above).

As for the other methods: I don't think it's in the scope of this PR to change them. And currently they only have debug checks.

Reply:
Sounds good. In that case I am on board with this change. Want to propose FCP?

Reply:
@dtolnay I'm not quite sure if you want me to add another sentence to this part of the documentation or not. (Will rebase either way soon.)

Regarding FCP: the main concern is [this one](https://github.com/rust-lang/rust/pull/66531#issuecomment-555782428). I replied to that comment, but no one else did, so I don't think it has been resolved yet. But maybe FCP is a way to get the others to look at this?